### ISSUE TITLE
[Open discussion] Flexible validator modules

### ISSUE THREAD

**Issue by tngan** [tngan/samlify#274]
Thanks for all feedback related to the latest implementation of validators injection. (Related issues #247 #249 #263 #265 #266)

From the release candidates of v2.5, we allowed users to specify their preferred validator modules via an environment variable `SAML_VALIDATOR`, but we removed that in the official release of v2.5 because using an env to configure the validator is not a good way to handle that kind of injection. 

Using `optionalDependencies` will fallback one by one if that particular module cannot be installed successfully, so it could probably install more than one modules in some systems but we never use it because for the runtime selection of validator to use is a trial-and-error process.

https://github.com/tngan/samlify/blob/89ad61092f8e1f2f9cb3972d07600274760e8347/src/schema-validator.ts#L26-L33

To keep this module simple and lightweight to use, I am planning to remove that `optionalDependencies` approach. **Instead, we won't do the installation for you. That means, if users install the validator that's listed in the whitelist, samlify can detect and use it. And if users doesn't install any validator, we would warn it for security concern.** 

The version range of the installed validator will be locked and specify in validator module.

**Comment by tngan** [comment]
I will leave this post opened about 2 weeks. See if we can have an even better solution.

**Comment by keerthivasan-r** [comment]
I have this question. Does `samlify` allows usage without having java installed in the machine? If yes, how to do that?

**Comment by tngan** [comment]
@keerthivasan-r Do you have npm error when you install samlify ? samlify itself doesn't require java installed, just one of the validator requires it. Now when you install samlify, it will install the validator dependencies (see in optionalDependencies) one by one, but it would ignore if some of it cannot be installed. Eventually, it would only pick the one that is installed to use during runtime.

**Comment by keerthivasan-r** [comment]
ok, I can understand that. but, i'm getting this strange issue referenced here.  it works in my local machine. but, it doesn't seem to work when containerized in a docker environment.

#204

**Comment by keerthivasan-r** [comment]
@tngan Do we have some docs, how this validator package loading works behind the scene? 

Even, the order of module resolution really matters? 
`const selectedValidator: string = moduleResolver(SchemaValidators.JAVAC) 
     || moduleResolver(SchemaValidators.LIBXML) 
     || moduleResolver(SchemaValidators.XMLLINT) 
     || moduleResolver(SchemaValidators.NATIVEXMLLINT) 
     || ''; ` 

Definitely, I would personally like to move away from java based dependency inside a node package. Do we not have other options in nodejs? I also see that we are installing open-jdk as part of Make build, it will not work, when we try to install inside docker-alphine images? 

 Your thoughts?

**Comment by tngan** [comment]
@keerthivasan-r I haven't updated the documentation yet. The order of module doesn't really matter if we leverage the installation of validator module to user side.

**Comment by keerthivasan-r** [comment]
@tngan I don't get it actually. Can you explain a bit more? What do you mean by 

> everage the installation of validator module to user side

**Comment by tngan** [comment]
@keerthivasan-r For tthe release v2.5, those validation modules will be installed one by one following the order specified in `optionalDependencies`, that's quite opinionated decision I have made before. There are some feedback after v2.5 is released, so I have started this thread and try to come up with a better solution regarding to those issues raised.

> leverage the installation of val.dator module to user side

It means users have to install the validator module themselves, samlify will only do version check for those whitelisted modules. If they don't install it, we will have run-time warning because of security concern. If they do install it, we will detect it based on the above-mentioned fallback mechanism.

**Comment by keerthivasan-r** [comment]
Sounds like a good one, we will have to remove them from `optionalDependencies` and leave it to users for installation of validators. What are the advantages of giving the users - flexibility with validator selection?

**Comment by tngan** [comment]
I think it would be better for samlify to do the interface instead of handling the plugin installation. So it is up to users to choose which one they are going to use, it is a common way to handle optional dependencies. 

I also got some feedbacks that they might not want the validation for some reasons, so it would be better to let users to make the decision.

**Comment by keerthivasan-r** [comment]
you mean that we can provide users to pass the validation function as part of samlify initialization parameters and that will do the job. Does that right?

**Comment by kiejo** [comment]
> To keep this module simple and lightweight to use, I am planning to remove that optionalDependencies approach. Instead, we won't do the installation for you.

This definitely sounds like an improvement to me!

Have you also thought about making the schema validator selection even more explicit, for example expecting the user to set a validator object? It could look something like this:
```javascript
const saml = require('samlify');
saml.setSchemaValidator({
  validate: (string) => {
    // Validator logic
  }
});
```

The existing schema validators could be moved into separate packages that directly depend on the different validator backends. Each validator package would have full control over which version of the backend to use removing an additional source of errors. The code could then look like this:
```javascript
const saml = require('samlify');
saml.setSchemaValidator(require('samlify-node-xmllint'));
```

This would remove any "magic" from the validator selection process and give the user full control over the validation logic. At the same time it would still be straightforward to just use a ready-made validator by explicitly requiring it in the code and adding it as a dependency.

**Comment by keerthivasan-r** [comment]
@kiejo Yeah, this will be the best thing to do away with all validation installation issues. That was my idea as well.

**Comment by tngan** [comment]
I also found out the reason why it throws error when installing the Java based validator module, because the `optionalDependencies` list contains a wrong package. Meanwhile, I will patch the module before I roll out the next minor version with new module installation.

**Comment by tngan** [comment]
@kiejo Nice idea, POC is made. b7c081c

You can also find the built package for the four validator modules under https://www.npmjs.com/search?q=authenio.

I will release a RC version of 2.6.0 in the coming week. Thanks for everyone here. Meanwhile, if you have any question or concern, don't be hesitate to discuss here.

**Comment by keerthivasan-r** [comment]
Thanks for the quick implementation and publishing the validator modules 👌

**Comment by keerthivasan-r** [comment]
> I also found out the reason why it throws error when installing the Java based validator module, because the `optionalDependencies` list contains a wrong package. Meanwhile, I will patch the module before I roll out the next minor version with new module installation.

You mean the package name or version is wrong or something??

**Comment by tngan** [comment]
@keerthivasan-r If you install latest version v2.5.0, the java based validator is not supposed to get installed because the `optionalDependencies` doesn't include the correct package. So, the chosen validator would not be the Java one based on the fallback approach.

https://github.com/tngan/samlify/blob/7bdfa423dc76a922d7a27df3ba0cdc67ddd70f91/package.json#L47-L49

It should be `@authenio/xsd-schema-validator` instead.

**Comment by matth-c3** [comment]
Hello, Any news on this by any chance ? 

Thanks

**Comment by tngan** [comment]
@matth-c3 Yes, please install 2.6.0-rc1 and checkout the installation guide.

https://github.com/tngan/samlify/tree/2.6.0#installation

**Comment by kiejo** [comment]
Thanks for the update! I tested the latest version with `@authenio/samlify-validate-with-xmllint` and `@authenio/samlify-node-xmllint` and so far did not encounter any blocking issues. Nice work :)

One small thing I noticed though is that I need to access the validator through a `default` property when using `require` instead of ES6 modules:
```javascript
const samlify = require('samlify');
const validator = require('@authenio/samlify-validate-with-xmllint').default;
samlify.setSchemaValidator(validator);
```
I think it would be nice if this was not necessary as it is somewhat unexpected for someone used to CommonJS modules, which I think is still the default in the Node.js ecosystem.
